rfc/0000-batch-cancellation.rst

```diff
@@ -0,0 +1,538 @@
+whether the batch has been cancelled, whether the job has been
+cancelled (failed parent dependencies), and whether the job is
+**always_run**. This trigger is currently ~170 lines and contains many
+nested if-else statements that are easy to make mistakes on when
+refactoring the code.
+
+When a batch is cancelled by either the user or the system
+("cancel_after_n_failures" or billing limits have been reached), the
+following occurs:
+
+1. A row for that batch id is inserted into a table that keeps track
+   of the batches that have been cancelled.
+2. The number of **cancellable** ready cores and number of ready jobs
+   for that batch are calculated by aggregating over the
+   ``batch_inst_coll_cancellable_resources`` table and then
+   subtracting those counts from the corresponding values in the
+   ``user_inst_coll_resources`` table.
+3. The rows pertaining to that batch id are deleted from the
+   ``batch_inst_coll_cancellable_resources`` table.
+4. Background canceller processes on the driver query the database for
+   jobs that are either Running, Creating, or Running and are in
+   cancelled batches and not **always_run** and then cancels them
+   individually. The canceller uses a slightly different fair share
+   algorithm to determine the number of jobs to cancel per user in a
+   fair share manner.
+
+As a historical note, the ``user_inst_coll_resources`` table that kept
+track of the n_ready_cores was a critical part of the autoscaler
+before changes in October 2022 were made to consider only the head of
+the job queue and compute the number of ready cores on the fly rather
+than from the aggregated global value. However, the fields in the
+``user_inst_coll_resources`` table are still used by the fair share
+algorithms, so we cannot remove these tables altogether.
+
+In summary, the current system is able to efficiently adapt to a
+cancellation event in O(1) operations, but the database structures
+that enable this efficiency are difficult to maintain due to the
+complexity of the trigger that does the count adjustments and the
+database structure is designed for only an entire batch to be
+cancelled.
+
+
+-----------------------------
+Proposed Change Specification
+-----------------------------
+
+**************
+Implementation
+**************
+
+Given the design of our current system, any changes to how
+cancellation works need to adhere to the following constraints:
+
+1. There must be a mechanism in the database that can be updated in an
+   O(1) database operation to indicate a batch has been cancelled.
+2. Queries that scan "Ready" jobs must be fast and respond to a
+   cancellation event instantaneously to avoid unnecessary
+   provisioning of resources.
+3. The data the fair share calculation uses must be updated
+   instantaneously after a cancellation event or be updated within
+   some time window that balances cluster efficiency with code
+   complexity or the fair share algorithm must be redesigned entirely.
```

Comment:
Constraint 1 is concise and precise. For 2 and 3, I think they should specify which queries exactly and what "fast" and "instantaneously" and "some time window" mean in both contexts.